File: apps/api/src/backfills.ts
import type { PipelineDefinition, Environment } from '@dataflow/shared';
import { pool, withTenant } from './db';
import { fireExecution } from './temporal';

const DAY_MS = 86_400_000;
const SUPPORTED_SOURCES = new Set(['postgres.fetch', 'mysql.fetch', 'mongodb.fetch']);

export interface BackfillPlan {
  from: string; to: string; partitionDays: number; maxConcurrency: number;
  partitionCount: number; estimatedExecutions: number;
  partitions: Array<{ from: string; to: string }>;
}

export function planBackfill(input: any): BackfillPlan {
  const fromMs = Date.parse(input?.from), toMs = Date.parse(input?.to);
  const partitionDays = Number(input?.partitionDays ?? 1);
  const maxConcurrency = Number(input?.maxConcurrency ?? 1);
  if (!Number.isFinite(fromMs) || !Number.isFinite(toMs) || fromMs >= toMs)
    throw new Error('from and to must be valid ISO timestamps with from before to');
  if (!Number.isInteger(partitionDays) || partitionDays < 1 || partitionDays > 31)
    throw new Error('partitionDays must be an integer between 1 and 31');
  if (!Number.isInteger(maxConcurrency) || maxConcurrency < 1 || maxConcurrency > 5)
    throw new Error('maxConcurrency must be an integer between 1 and 5');
  const partitions = [];
  for (let cursor = fromMs; cursor < toMs; cursor += partitionDays * DAY_MS) {
    partitions.push({
      from: new Date(cursor).toISOString(),
      to: new Date(Math.min(toMs, cursor + partitionDays * DAY_MS)).toISOString(),
    });
    if (partitions.length > 366) throw new Error('backfill exceeds the 366 partition limit');
  }
  return {
    from: new Date(fromMs).toISOString(), to: new Date(toMs).toISOString(),
    partitionDays, maxConcurrency, partitionCount: partitions.length,
    estimatedExecutions: partitions.length, partitions,
  };
}

export function validateBackfillSources(definition: PipelineDefinition): void {
  const sources = definition.nodes.filter(node => node.type === 'source');
  if (!sources.length) throw new Error('pipeline has no source');
  for (const source of sources) {
    if (!SUPPORTED_SOURCES.has(source.activityType))
      throw new Error(`${source.activityType}: partitioned backfill is supported only for PostgreSQL, MySQL, and MongoDB cursor sources`);
    if (source.config?.syncMode === 'cdc')
      throw new Error(`${source.activityType}: CDC sources cannot run partitioned backfills; switch to cursor mode`);
    if (source.config?.syncMode !== 'cursor')
      throw new Error(`${source.activityType}: partitioned backfills require cursor mode`);
    if (source.config?.cursorType !== 'date')
      throw new Error(`${source.activityType}: partitioned backfills require cursorType=date`);
    const cursorField = source.activityType === 'mongodb.fetch'
      ? source.config?.cursorField : source.config?.cursorColumn;
    if (!cursorField) throw new Error(`${source.activityType}: partitioned backfills require a cursor field`);
  }
}

export async function dispatchBackfillPartitions(limit = 20): Promise<number> {
  let started = 0;
  for (; started < limit; started++) {
    const { rows } = await pool.query(`SELECT * FROM claim_next_backfill_partition()`);
    const claimed = rows[0];
    if (!claimed) break;
    const definition = structuredClone(claimed.definition) as PipelineDefinition;
    for (const node of definition.nodes.filter(node => node.type === 'source')) {
      node.ingestion = {
        ...node.ingestion, mode: 'backfill',
        backfillStart: new Date(claimed.range_start).toISOString(),
        backfillEnd: new Date(claimed.range_end).toISOString(),
        stateKey: claimed.partition_id,
        pageSize: Math.max(Number(node.ingestion?.pageSize ?? 0), 10_000),
      };
    }
    try {
      await fireExecution(
        definition, claimed.pipeline_id, 'backfill', claimed.environment as Environment,
        undefined, undefined, undefined, claimed.partition_id,
      );
    } catch (error: any) {
      await withTenant(claimed.tenant_id, client => client.query(
        `WITH changed AS (
           UPDATE backfill_partitions SET status='failed',error=$2,completed_at=now() WHERE id=$1 RETURNING job_id
         )
         UPDATE backfill_jobs bj SET status='failed',completed_at=now()
           FROM changed WHERE bj.id=changed.job_id
            AND NOT EXISTS (SELECT 1 FROM backfill_partitions p
                              WHERE p.job_id=bj.id AND p.status IN ('pending','starting','running'))`,
        [claimed.partition_id, String(error?.message ?? error).slice(0, 2000)],
      ));
    }
  }
  return started;
}

export function startBackfillDispatcher(): void {
  const tick = async () => {
    try { await dispatchBackfillPartitions(); }
    catch (error: any) { console.error('backfill dispatcher failed', error.message); }
  };
  void tick();
  setInterval(tick, Number(process.env.BACKFILL_DISPATCH_INTERVAL_MS ?? 2000)).unref();
}

File: apps/workflow-go/internal/api/backfills.go
package api

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"math"
	"time"

	"github.com/dataflow-poc/workflow-go/internal/model"
	"github.com/jackc/pgx/v5"
)

type backfillPlan struct {
	From                string              `json:"from"`
	To                  string              `json:"to"`
	PartitionDays       int                 `json:"partitionDays"`
	MaxConcurrency      int                 `json:"maxConcurrency"`
	PartitionCount      int                 `json:"partitionCount"`
	EstimatedExecutions int                 `json:"estimatedExecutions"`
	Partitions          []map[string]string `json:"partitions"`
}

func planBackfill(input map[string]interface{}) (backfillPlan, error) {
	from, fromErr := time.Parse(time.RFC3339, stringValue(input["from"]))
	to, toErr := time.Parse(time.RFC3339, stringValue(input["to"]))
	if fromErr != nil || toErr != nil || !from.Before(to) {
		return backfillPlan{}, fmt.Errorf("from and to must be valid ISO timestamps with from before to")
	}
	partitionDays := int(numberValue(input["partitionDays"], 1))
	maxConcurrency := int(numberValue(input["maxConcurrency"], 1))
	if partitionDays < 1 || partitionDays > 31 {
		return backfillPlan{}, fmt.Errorf("partitionDays must be an integer between 1 and 31")
	}
	if maxConcurrency < 1 || maxConcurrency > 5 {
		return backfillPlan{}, fmt.Errorf("maxConcurrency must be an integer between 1 and 5")
	}
	partitions := []map[string]string{}
	for cursor := from; cursor.Before(to); cursor = cursor.AddDate(0, 0, partitionDays) {
		end := cursor.AddDate(0, 0, partitionDays)
		if end.After(to) {
			end = to
		}
		partitions = append(partitions, map[string]string{"from": cursor.UTC().Format(time.RFC3339Nano), "to": end.UTC().Format(time.RFC3339Nano)})
		if len(partitions) > 366 {
			return backfillPlan{}, fmt.Errorf("backfill exceeds the 366 partition limit")
		}
	}
	return backfillPlan{From: from.UTC().Format(time.RFC3339Nano), To: to.UTC().Format(time.RFC3339Nano), PartitionDays: partitionDays,
		MaxConcurrency: maxConcurrency, PartitionCount: len(partitions), EstimatedExecutions: len(partitions), Partitions: partitions}, nil
}

func validateBackfillSources(def model.PipelineDefinition) error {
	supported := map[string]bool{"postgres.fetch": true, "mysql.fetch": true, "mongodb.fetch": true}
	found := false
	for _, node := range def.Nodes {
		if node.Type != "source" {
			continue
		}
		found = true
		if !supported[node.ActivityType] {
			return fmt.Errorf("%s: partitioned backfill is supported only for PostgreSQL, MySQL, and MongoDB cursor sources", node.ActivityType)
		}
		if node.Config["syncMode"] != "cursor" {
			return fmt.Errorf("%s: partitioned backfills require cursor mode", node.ActivityType)
		}
		if node.Config["cursorType"] != "date" {
			return fmt.Errorf("%s: partitioned backfills require cursorType=date", node.ActivityType)
		}
		field := node.Config["cursorColumn"]
		if node.ActivityType == "mongodb.fetch" {
			field = node.Config["cursorField"]
		}
		if stringValue(field) == "" {
			return fmt.Errorf("%s: partitioned backfills require a cursor field", node.ActivityType)
		}
	}
	if !found {
		return fmt.Errorf("pipeline has no source")
	}
	return nil
}

func (s *Server) backfillDispatcher(ctx context.Context) {
	ticker := time.NewTicker(s.Config.BackfillDispatchInterval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			if _, err := s.dispatchBackfills(ctx, 20); err != nil {
				slog.Error("backfill dispatcher failed", "error", err)
			}
		}
	}
}

func (s *Server) dispatchBackfills(ctx context.Context, limit int) (int, error) {
	started := 0
	for ; started < limit; started++ {
		rows, err := s.DB.Pool.Query(ctx, `SELECT * FROM claim_next_backfill_partition()`)
		if err != nil {
			return started, err
		}
		claimed, err := oneMap(rows)
		if err != nil || claimed == nil {
			return started, err
		}
		body, _ := json.Marshal(claimed["definition"])
		var def model.PipelineDefinition
		if err := json.Unmarshal(body, &def); err != nil {
			return started, err
		}
		for i := range def.Nodes {
			node := &def.Nodes[i]
			if node.Type != "source" {
				continue
			}
			if node.Ingestion == nil {
				node.Ingestion = &model.IngestionConfig{}
			}
			node.Ingestion.Mode = "backfill"
			node.Ingestion.BackfillStart = timeValue(claimed["range_start"]).Format(time.RFC3339Nano)
			node.Ingestion.BackfillEnd = timeValue(claimed["range_end"]).Format(time.RFC3339Nano)
			node.Ingestion.StateKey = stringValue(claimed["partition_id"])
			node.Ingestion.PageSize = int(math.Max(float64(node.Ingestion.PageSize), 10000))
		}
		_, err = s.fireExecution(ctx, def, stringValue(claimed["pipeline_id"]), "backfill", model.Environment(stringValue(claimed["environment"])), nil, "", "", stringValue(claimed["partition_id"]))
		if err != nil {
			_ = s.DB.TenantTx(ctx, stringValue(claimed["tenant_id"]), func(tx pgx.Tx) error {
				_, updateErr := tx.Exec(ctx, `UPDATE backfill_partitions SET status='failed',error=$2,completed_at=now() WHERE id=$1`, claimed["partition_id"], err.Error())
				return updateErr
			})
		}
	}
	return started, nil
}

func numberValue(value interface{}, fallback float64) float64 {
	if n, ok := value.(float64); ok {
		return n
	}
	return fallback
}
func timeValue(value interface{}) time.Time {
	if t, ok := value.(time.Time); ok {
		return t
	}
	t, _ := time.Parse(time.RFC3339, stringValue(value))
	return t
}
